sns_ik_lib/src/sns_vel_ik_base.cpp

```diff
@@ -0,0 +1,393 @@
+    } else {
+      return 1.0;  // feasible without scaling!
+    }
+  } else {
+    return 0.0;  // infeasible
+  }
+}
+
+/*************************************************************************************************
+ *                                 Public Methods                                                *
+ *************************************************************************************************/
+
+SnsVelIkBase::uPtr SnsVelIkBase::create(int nJnt)
+{
+  if (nJnt <= 0) {
+    ROS_ERROR("Bad Input: dqLow.size(%d) > 0 is required!", nJnt);
+    return nullptr;
+  }
+  Eigen::ArrayXd dqLow = NEG_INF*Eigen::ArrayXd::Ones(nJnt);
+  Eigen::ArrayXd dqUpp = POS_INF*Eigen::ArrayXd::Ones(nJnt);
+  return create(dqLow, dqUpp);
+}
+
+/*************************************************************************************************/
+
+SnsVelIkBase::uPtr SnsVelIkBase::create(const Eigen::ArrayXd& dqLow, const Eigen::ArrayXd& dqUpp)
+{
+  // Input validation
+  int nJnt = dqLow.size();
+  if (nJnt <= 0) {
+    ROS_ERROR("Bad Input: dqLow.size(%d) > 0 is required!", nJnt);
+    return nullptr;
+  }
+
+  // Create an empty solver
+  SnsVelIkBase::uPtr velIk(new SnsVelIkBase(nJnt));
+
+  // Set the joint limits:
+  if (!velIk->setVelBnd(dqLow, dqUpp)) { ROS_ERROR("Bad Input!"); return nullptr; };
+
+  return velIk;
+}
+
+/*************************************************************************************************/
+
+bool SnsVelIkBase::setVelBnd(const Eigen::ArrayXd& dqLow, const Eigen::ArrayXd& dqUpp)
+{
+  int nJnt = dqLow.size();
+  if (nJnt <= 0) {
+    ROS_ERROR("Bad Input: dqLow.size(%d) > 0 is required!", nJnt);
+    return false;
+  }
+  if (dqLow.size() != dqUpp.size()) {
+    ROS_ERROR("Bad Input: dqLow.size(%d) == dqUpp.size(%d) is required!",
+              int(dqLow.size()), int(dqUpp.size()));
+    return false;
+  }
+  nJnt_ = nJnt;
+  dqLow_ = dqLow;
+  dqUpp_ = dqUpp;
+  return true;
+}
+
+/*************************************************************************************************/
+
+SnsIkExitCode SnsVelIkBase::solve(const Eigen::MatrixXd& J, const Eigen::VectorXd& dx,
+                               Eigen::VectorXd* dq, double* taskScale)
+{
+  // Input validation
+  if (!dq) { ROS_ERROR("dq is nullptr!"); return SnsIkExitCode::BadUserInput; }
+  if (!taskScale) { ROS_ERROR("taskScale is nullptr!"); return SnsIkExitCode::BadUserInput; }
+  int nTask = dx.size();
+  if (nTask <= 0) {
+    ROS_ERROR("Bad Input: dx.size() > 0 is required!");
+    return SnsIkExitCode::BadUserInput;
+  }
+  if (int(J.rows()) != nTask) {
+    ROS_ERROR("Bad Input: J.rows() == dx.size() is required!");
+    return SnsIkExitCode::BadUserInput;
+  }
+  if (int(J.cols()) != nJnt_) {
+    ROS_ERROR("Bad Input: J.cols() == nJnt is required!");
+    return SnsIkExitCode::BadUserInput;
+  }
+
+  // Local variable initialization:
+  Eigen::MatrixXd W = Eigen::MatrixXd::Identity(nJnt_, nJnt_);  // null-space selection matrix
```

Comment:
Let's give more description to W. W is a diagonal selection matrix which indicates enabled joints. The entry of 1 indicates the corresponding joint is enabled for the task, and 0 indicates the corresponding joint is saturated.

Reply:
Good description - I added it to the comments.